### ISSUE TITLE
GoRouter doesn't notify when location changed to same routes with different params

### ISSUE THREAD

**Issue by SuperPenguin** [flutter/flutter#108674]
## Steps to Reproduce

1. Run example code
2. Click on `/family/f1/person/p1` button, see log output
3. Click on `/family/f1/person/p2` button, see log output

**Expected results:** go router should notify on location changed

**Actual results:** go router doesn't notify

<details>
<summary>Code sample</summary>

```dart
import 'package:flutter/foundation.dart';
import 'package:flutter/material.dart';
import 'package:go_router/go_router.dart';

void main() {
  runApp(const App());
}

class App extends StatefulWidget {
  const App({Key? key}) : super(key: key);

  @override
  State<App> createState() => _AppState();
}

class _AppState extends State<App> {
  final _goRouter = GoRouter(
    routes: [
      GoRoute(
        path: '/',
        builder: (context, state) => const HomeScreen(),
        routes: [
          GoRoute(
            path: 'family/:fid',
            builder: (context, state) {
              final family = state.params['fid'] as String;
              return FamilyScreen(family: family);
            },
            routes: [
              GoRoute(
                path: 'person/:pid',
                builder: (context, state) {
                  final person = state.params['pid'] as String;
                  return PersonScreen(person: person);
                },
              ),
            ],
          ),
        ],
      ),
    ],
  );

  void _onNotify() {
    if (kDebugMode) {
      print('[GoRouter] ${_goRouter.location}');
    }
  }

  @override
  void initState() {
    super.initState();
    _goRouter.addListener(_onNotify);
  }

  @override
  void dispose() {
    _goRouter.removeListener(_onNotify);
    _goRouter.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp.router(
      routeInformationParser: _goRouter.routeInformationParser,
      routerDelegate: _goRouter.routerDelegate,
      routeInformationProvider: _goRouter.routeInformationProvider,
    );
  }
}

class HomeScreen extends StatelessWidget {
  const HomeScreen({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text('Home'),
      ),
      body: const SharedBody(),
    );
  }
}

class FamilyScreen extends StatelessWidget {
  const FamilyScreen({
    Key? key,
    required this.family,
  }) : super(key: key);

  final String family;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('Family $family'),
      ),
      body: const SharedBody(),
    );
  }
}

class PersonScreen extends StatelessWidget {
  const PersonScreen({
    Key? key,
    required this.person,
  }) : super(key: key);

  final String person;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('Person $person'),
      ),
      body: const SharedBody(),
    );
  }
}

class SharedBody extends StatelessWidget {
  const SharedBody({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    const List<String> routes = [
      '/',
      '/family/f1',
      '/family/f1/person/p1',
      '/family/f1/person/p2',
    ];

    return Center(
      child: Column(
        mainAxisSize: MainAxisSize.min,
        children: [
          for (final r in routes)
            ElevatedButton(
              onPressed: () => GoRouter.of(context).go(r),
              child: Text(r),
            ),
        ],
      ),
    );
  }
}
```

```
[✓] Flutter (Channel stable, 3.0.5, on macOS 12.4 21F79 darwin-arm, locale en-ID)
    • Flutter version 3.0.5 at /Users/s21/Documents/GitHub/flutter/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision f1875d570e (2 weeks ago), 2022-07-13 11:24:16 -0700
    • Engine revision e85ea0e79c
    • Dart version 2.17.6
    • DevTools version 2.12.2

[✓] Android toolchain - develop for Android devices (Android SDK version 31.0.0)
    • Android SDK at /Users/s21/Library/Android/sdk
    • Platform android-33, build-tools 31.0.0
    • Java binary at: /Applications/Android Studio.app/Contents/jre/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.12+0-b1504.28-7817840)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 13.4.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • CocoaPods version 1.11.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2021.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.12+0-b1504.28-7817840)

[✓] VS Code (version 1.69.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.44.0

[✓] Connected device (4 available)
    • SM G991B (mobile)        • RRCR2003SSA               • android-arm64  • Android 12 (API 31)
    • Andree’s iPhone (mobile) • 00008030-000314393443802E • ios            • iOS 15.6 19G71
    • macOS (desktop)          • macos                     • darwin-arm64   • macOS 12.4 21F79 darwin-arm
    • Chrome (web)             • chrome                    • web-javascript • Google Chrome 103.0.5060.134
    ! Error: Andree’s iPhone is busy: Fetching debug symbols for Andree’s iPhone. Xcode will continue when Andree’s iPhone is finished. (code -10)

[✓] HTTP Host Availability
    • All required HTTP hosts are available

• No issues found!
```

</details>

GoRouter version 4.2.2

**Comment by maheshmnj** [comment]
Hi @SuperPenguin, Thanks for filing the issue. I am able to reproduce the issue on the latest stable and the master channel with ` go_router: ^4.2.2` 

 
<details>
<summary>code sample</summary>

```dart
import 'package:flutter/foundation.dart';
import 'package:flutter/material.dart';
import 'package:go_router/go_router.dart';

void main() {
  runApp(const App());
}

class App extends StatefulWidget {
  const App({Key? key}) : super(key: key);

  @override
  State<App> createState() => _AppState();
}

class _AppState extends State<App> {
  final _goRouter = GoRouter(
    routes: [
      GoRoute(
        path: '/',
        builder: (context, state) => const HomeScreen(),
        routes: [
          GoRoute(
            path: 'family/:fid',
            builder: (context, state) {
              final family = state.params['fid'] as String;
              return FamilyScreen(family: family);
            },
            routes: [
              GoRoute(
                path: 'person/:pid',
                builder: (context, state) {
                  final person = state.params['pid'] as String;
                  return PersonScreen(person: person);
                },
              ),
            ],
          ),
        ],
      ),
    ],
  );

  void _onNotify() {
    if (kDebugMode) {
      print('[GoRouter] ${_goRouter.location}');
    }
  }

  @override
  void initState() {
    super.initState();
    _goRouter.addListener(_onNotify);
  }

  @override
  void dispose() {
    _goRouter.removeListener(_onNotify);
    _goRouter.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp.router(
      routeInformationParser: _goRouter.routeInformationParser,
      routerDelegate: _goRouter.routerDelegate,
      routeInformationProvider: _goRouter.routeInformationProvider,
    );
  }
}

class HomeScreen extends StatelessWidget {
  const HomeScreen({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text('Home'),
      ),
      body: const SharedBody(),
    );
  }
}

class FamilyScreen extends StatelessWidget {
  const FamilyScreen({
    Key? key,
    required this.family,
  }) : super(key: key);

  final String family;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('Family $family'),
      ),
      body: const SharedBody(),
    );
  }
}

class PersonScreen extends StatelessWidget {
  const PersonScreen({
    Key? key,
    required this.person,
  }) : super(key: key);

  final String person;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('Person $person'),
      ),
      body: const SharedBody(),
    );
  }
}

class SharedBody extends StatelessWidget {
  const SharedBody({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    const List<String> routes = [
      '/',
      '/family/f1',
      '/family/f1/person/p1',
      '/family/f1/person/p2',
    ];

    return Center(
      child: Column(
        mainAxisSize: MainAxisSize.min,
        children: [
          for (final r in routes)
            Padding(
              padding: const EdgeInsets.all(8.0),
              child: ElevatedButton(
                onPressed: () => GoRouter.of(context).go(r),
                child: Text(r),
              ),
            ),
        ],
      ),
    );
  }
}

```
</details> 

<details>
<summary>logs</summary>

```
mahesh@Maheshs-MacBook-Air-M1 sample % flutter run -d chrome
Running "flutter pub get" in sample...                           1,481ms
Launching lib/main.dart on Chrome in debug mode...
Waiting for connection from debug service on Chrome...              9.8s
This app is linked to the debug service: ws://127.0.0.1:63122/DC7K52AQ9l4=/ws
Debug service listening on ws://127.0.0.1:63122/DC7K52AQ9l4=/ws

💪 Running with sound null safety 💪

🔥  To hot restart changes while running, press "r" or "R".
For a more detailed help message, press "h". To quit, press "q".

An Observatory debugger and profiler on Chrome is available at: http://127.0.0.1:63122/DC7K52AQ9l4=
Flutter Web Bootstrap: Programmatic
[GoRouter] /
The Flutter DevTools debugger and profiler on Chrome is available at:
http://127.0.0.1:9101?uri=http://127.0.0.1:63122/DC7K52AQ9l4=

Performing hot restart...                                        1,303ms
Restarted application in 1,303ms.
Flutter Web Bootstrap: Programmatic
[GoRouter] /
[GoRouter] /family/f1
[GoRouter] /family/f1/person/p1
```
</details> 

<details>
<summary>flutter doctor -v (mac)</summary>

```
[✓] Flutter (Channel stable, 3.0.5, on macOS 12.4 21F79 darwin-arm, locale en-IN)
    • Flutter version 3.0.5 at /Users/mahesh/Documents/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision f1875d570e (5 days ago), 2022-07-13 11:24:16 -0700
    • Engine revision e85ea0e79c
    • Dart version 2.17.6
    • DevTools version 2.12.2

[✓] Android toolchain - develop for Android devices (Android SDK version 33.0.0-rc4)
    • Android SDK at /Users/mahesh/Library/Android/sdk
    • Platform android-32, build-tools 33.0.0-rc4
    • ANDROID_HOME = /Users/mahesh/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jre/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.12+0-b1504.28-7817840)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 13.2.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • CocoaPods version 1.11.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2021.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.12+0-b1504.28-7817840)

[✓] IntelliJ IDEA Community Edition (version 2021.2.1)
    • IntelliJ at /Applications/IntelliJ IDEA CE.app
    • Flutter plugin version 61.2.4
    • Dart plugin version 212.5080.8

[✓] VS Code (version 1.67.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.44.0

[✓] Connected device (3 available)
    • sdk gphone arm64 (mobile) • emulator-5554 • android-arm64  • Android 11 (API 30) (emulator)
    • macOS (desktop)           • macos         • darwin-arm64   • macOS 12.4 21F79 darwin-arm
    • Chrome (web)              • chrome        • web-javascript • Google Chrome 103.0.5060.114

[✓] HTTP Host Availability
    • All required HTTP hosts are available

• No issues found!
```

```
[✓] Flutter (Channel master, 3.1.0-0.0.pre.1868, on macOS 12.4 21F79 darwin-arm, locale en-IN)
    • Flutter version 3.1.0-0.0.pre.1868 on channel master at /Users/mahesh/Documents/flutter_master
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 925bee92e7 (6 days ago), 2022-07-26 04:49:06 -0400
    • Engine revision 5dcaeae656
    • Dart version 2.19.0 (build 2.19.0-34.0.dev)
    • DevTools version 2.15.0

[✓] Android toolchain - develop for Android devices (Android SDK version 33.0.0-rc4)
    • Android SDK at /Users/mahesh/Library/Android/sdk
    • Platform android-32, build-tools 33.0.0-rc4
    • ANDROID_HOME = /Users/mahesh/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jre/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.12+0-b1504.28-7817840)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 13.2.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 13C100
    • CocoaPods version 1.11.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2021.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.12+0-b1504.28-7817840)

[✓] IntelliJ IDEA Community Edition (version 2021.2.1)
    • IntelliJ at /Applications/IntelliJ IDEA CE.app
    • Flutter plugin version 61.2.4
    • Dart plugin version 212.5080.8

[✓] VS Code (version 1.69.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.44.0

[✓] Connected device (3 available)
    • sdk gphone arm64 (mobile) • emulator-5554 • android-arm64  • Android 11 (API 30) (emulator)
    • macOS (desktop)           • macos         • darwin-arm64   • macOS 12.4 21F79 darwin-arm
    • Chrome (web)              • chrome        • web-javascript • Google Chrome 103.0.5060.134

[✓] HTTP Host Availability
    • All required HTTP hosts are available

• No issues found!
```
</details>

**Comment by maheshmnj** [comment]
cc: @chunhtai

**Comment by cgestes** [comment]
to get all notify:
```
     router.routeInformationProvider.addListener(listen);
     router.routerDelegate.addListener(listen);
     router.addListener(listen);
```

I suspect it is an internal API.

**Comment by chunhtai** [comment]
This should be fixed in 5.1.2
